src/renderer/state.ts

```diff
@@ -329,41 +351,110 @@ export class AppState {
     );
     window.ElectronFiddle.addEventListener('before-quit', this.setIsQuitting);
 
+    /**
+     * Listens for changes in the app settings made in other windows
+     * and refreshes the current window settings accordingly.
+     */
+    window.addEventListener('storage', (event) => {
+      const key = event.key as GlobalSettingKey;
+      const { newValue } = event;
+
+      let parsedValue: unknown;
+
+      try {
+        parsedValue = JSON.parse(newValue as string) as unknown;
+      } catch {
+        // The new value is a plain string, not a well-formed stringified object.
+        parsedValue = newValue;
+      }
+
+      if (key in GlobalSetting && key in this) {
+        // Any settings listed here need morethan just updating the state directly.
+        const actions: Partial<Record<GlobalSettingKey, () => void>> = {
+          theme: () => {
+            this.setTheme(parsedValue as string);
+          },
+        };
+
+        const action = actions[key];
+
+        if (typeof action === 'function') {
+          action();
+        } else {
+          // Fall back to updating the state.
+          this[key] = parsedValue;
+        }
```

Comment:
The idea of using an enum for keys was to leverage exhaustiveness checking so that if a new setting is added, there will be a TS error until the key is handled here, to prevent missing a needed action (like `setTheme`). It's more verbose, but makes it more difficult to forget to handle it properly.

So originally I was thinking something like:

```js
switch(key) {
  case GlobalSetting.theme:
    this.setTheme(parsedValue as string);
    break;
  case GlobalSetting.fontFamily:
  case GlobalSetting.fontSize:
    this[key] = parsedValue;
}
```

Thoughts on that approach versus the one you've implemented here?

Reply:
I missed the bit about having a compile-time error when a new setting is added, it makes total sense. We could even do that using my object approach by doing something like this...

```ts
const actions: Record<GlobalSettingKey, null | (() => void)> = {
  acceleratorsToBlock: null,
  channelsToShow: null,
  electronMirror: null,
  environmentVariables: null,
  executionFlags: null,
  fontFamily: null,
  fontSize: null,
  gitHubAvatarUrl: null,
  gitHubLogin: null,
  gitHubName: null,
  gitHubToken: null,
  hasShownTour: null,
  isClearingConsoleOnRun: null,
  isEnablingElectronLogging: null,
  isKeepingUserDataDirs: null,
  isPublishingGistAsRevision: null,
  isUsingSystemTheme: null,
  knownVersion: null,
  localVersion: null,
  packageAuthor: null,
  packageManager: null,
  showObsoleteVersions: null,
  showUndownloadedVersions: null,
  theme: () => {
    this.setTheme(parsedValue as string);
  },
};
```

...but that would still require that `action` validation, so I think using a `switch` with a type guard looks cleaner:

```ts
switch (key) {
  case 'theme': {
    this.setTheme(parsedValue as string);
    break;
  }

  case 'acceleratorsToBlock':
  case 'channelsToShow':
  case 'electronMirror':
  case 'environmentVariables':
  case 'executionFlags':
  case 'fontFamily':
  case 'fontSize':
  case 'gitHubAvatarUrl':
  case 'gitHubLogin':
  case 'gitHubName':
  case 'gitHubToken':
  case 'hasShownTour':
  case 'isClearingConsoleOnRun':
  case 'isEnablingElectronLogging':
  case 'isKeepingUserDataDirs':
  case 'isPublishingGistAsRevision':
  case 'isUsingSystemTheme':
  case 'knownVersion':
  case 'localVersion':
  case 'packageAuthor':
  case 'packageManager':
  case 'showObsoleteVersions':
  case 'showUndownloadedVersions': {
    // Fall back to updating the state.
    this[key] = parsedValue;
    break;
  }

  default: {
    settingKeyTypeGuard(key);
  }
}

// ...

function settingKeyTypeGuard(key: never): never {
  throw new Error(
    `Unhandled setting ${key}, please handle it in the \`AppState\`.`,
  );
}
```

Reply:
Sounds good to me. 👍